[namedtensor] Added at::Dimname

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #21373 [namedtensor] Add named_guard to native_functions options
* #21341 [namedtensor] Add unique_ptr<NamedTensorMeta> field to TensorImpl
* #21281 [namedtensor] Implement at::match(Dimname, Dimname) and at::unify(Dimname, Dimname)
* **#21280 [namedtensor] Added at::Dimname**

Dimname represents a dimension name for named tensors. Because we
support special names (wildcard, tagged, and possibly more in the
future) a single at::Symbol does not suffice and we need this struct.

Test Plan:
- build/bin/Dimname_test

gh-metadata: pytorch pytorch 21280 gh/zou3519/39/head

Differential Revision: [D15698516](https://our.internmc.facebook.com/intern/diff/D15698516)

zou3519 committed:
[namedtensor] Added at::Dimname

Dimname represents a dimension name for named tensors. Because we
support special names (wildcard, tagged, and possibly more in the
future) a single at::Symbol does not suffice and we need this struct.

Test Plan:
- build/bin/Dimname_test

gchanan commented on aten/src/ATen/core/Dimname_test.cpp:
```diff
@@ -0,0 +1,57 @@
+#ifdef NAMEDTENSOR_ENABLED
```
do we ever build this?

zou3519 replied:
Yes, it always gets built

gchanan replied:
so what happens if you try to run this, then?

zou3519 replied:
if NAMEDTENSOR_ENABLED, then
```
Running main() from ../third_party/googletest/googletest/src/gtest_main.cc
[==========] Running 5 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 5 tests from DimnameTest
[ RUN      ] DimnameTest.isValidIdentifier
[       OK ] DimnameTest.isValidIdentifier (0 ms)
[ RUN      ] DimnameTest.wildcardName
[       OK ] DimnameTest.wildcardName (0 ms)
[ RUN      ] DimnameTest.createNormalName
[       OK ] DimnameTest.createNormalName (2 ms)
[ RUN      ] DimnameTest.createTaggedName
[       OK ] DimnameTest.createTaggedName (3 ms)
[ RUN      ] DimnameTest.unifyAndMatch
[       OK ] DimnameTest.unifyAndMatch (0 ms)
[----------] 5 tests from DimnameTest (5 ms total)

[----------] Global test environment tear-down
[==========] 5 tests from 1 test case ran. (5 ms total)
[  PASSED  ] 5 tests.
```
otherwise there are 0 tests.

zou3519 replied:
I should figure out how to hook this up to the CI, I'm not sure this is being run right now

zou3519 committed:
Update on "[namedtensor] Added at::Dimname"

[namedtensor] Added at::Dimname

Dimname represents a dimension name for named tensors. Because we
support special names (wildcard, tagged, and possibly more in the
future) a single at::Symbol does not suffice and we need this struct.

Test Plan:
- build/bin/Dimname_test

gh-metadata: pytorch pytorch 21280 gh/zou3519/39/head

zou3519 committed:
Update on "[namedtensor] Added at::Dimname"

[namedtensor] Added at::Dimname

Dimname represents a dimension name for named tensors. Because we
support special names (wildcard, tagged, and possibly more in the
future) a single at::Symbol does not suffice and we need this struct.

Test Plan:
- build/bin/Dimname_test

gh-metadata: pytorch pytorch 21280 gh/zou3519/39/head

zou3519 committed:
Update on "[namedtensor] Added at::Dimname"

[namedtensor] Added at::Dimname

Dimname represents a dimension name for named tensors. Because we
support special names (wildcard, tagged, and possibly more in the
future) a single at::Symbol does not suffice and we need this struct.

Test Plan:
- build/bin/Dimname_test

gh-metadata: pytorch pytorch 21280 gh/zou3519/39/head
